sdk/core/core-client/src/serviceClient.ts

```diff
@@ -286,28 +285,15 @@ function flattenResponse(
   fullResponse: FullOperationResponse,
   responseSpec: OperationResponseMap | undefined
 ): OperationResponse {
-  const parsedHeaders = fullResponse.parsedHeaders;
   const bodyMapper = responseSpec && responseSpec.bodyMapper;
 
-  function addResponse<T extends object>(
-    obj: T
-  ): T & { readonly _response: FullOperationResponse } {
-    return Object.defineProperty(obj, "_response", {
-      configurable: false,
-      enumerable: false,
-      writable: false,
-      value: fullResponse
-    });
-  }
-
   if (bodyMapper) {
     const typeName = bodyMapper.type.name;
     if (typeName === "Stream") {
-      return addResponse({
-        ...parsedHeaders,
+      return {
```

Comment:
why did we get rid of the parsedHeaders in some of these other cases? I thought there are times when swagger has things in headers that we want to put on the body of what we return from the client, such as `etag`

Reply:
I am not really sure, I have no experience with these type of responses. My default is to not include headers so I went with that for streams and arrays. I will investigate further but in the meantime, please feel free to share examples of why this current behavior will not work.

Reply:
Streams could be like a blob download, right? And that could still need header data on the object besides the blob body to grab the actual bytes. Since the download is streaming the only way to include metadata would be through headers.

Similar to arrays, if the response body is [1, 2, 3, 4] but the service wants to include some metadata about the response (e.g. relevance metrics) then headers let us attach extra properties to the array itself, since arrays are objects too.

Reply:
I understand headers are important and this work still makes them available albeit through one extra hoop. I think we should be consistent and make all headers in all situations available only through the callback. To be more specific, my thinking is:
- we do not want to attach anything (even headers) to the value represented in the response body because it might be null/undefined.
- there can be situations where we know there can not be a null (streams?) but we still want to be consistent and make all headers available through the callback only.

I can be wrong though so I am happy to discuss this further offline when you are back.

Reply:
This is really interesting and we should grab someone who is more familiar with the intent of OpenAPI. (/cc @bterlson @chradek )

As someone making a RESTful endpoint, I don't necessarily consider the body to be the only place that information is returned. For example, sometimes you can make a request where the only interesting information is control metadata in the headers. Rather than having to expect the client to return `null` or `undefined` and use the raw response mechanism to fish out the headers, I imagine returning an object like `{ header1: value, header2: value }` would be more useful.

The most ambiguous case I can think of is when the body is a simple type like number/string such as in this example:

http://spec.openapis.org/oas/v3.0.3#response-object-examples

My take on this is that if header values are important enough to document in swagger as part of the response, you should attach them to the default result. For simple content types like the above, I think a wrapping object (e.g. `{ value: "woah", "X-Rate-Limit-Limit": 10 }`) would make sense, since that provides everything described in swagger to the caller. 

Since this is the most expressive behavior, I think it should be the default, and we could always re-simplify in convenience layers or do swagger transforms. 

That said, if there were no described headers in the response, then I think the string should be returned directly.

Reply:
I'm not sure how the intent of OpenAPI is useful here? From my perspective, though, users of our libraries shouldn't have to care whether a service is serializing something over HTTP using headers or inside the body somehow (there are often good reasons for either and sometimes both) for anything but obscure use cases.

So say for example a service uses the etag header but does not have etag in the JSON representation in the response body. It would be pretty frustrating for users to fish out etag using the callback, so I'd expect it to be added to the return value somehow. Something like rate limit headers or retry-after headers are more like things I could see using the callback for if users need them since presumably our policies are handling them for the user.

Reply:
I agree with @bterlson on this.

Does this change make it so we can no longer access fields that are populated from headers even if they are modeled in OpenAPI? That would be very annoying as a user and API author.

Reply:
This PR as currently written removes fields populated from headers for all requests except for `HEAD` requests.

Reply:
Ah, I don't think that's the right direction. That would make using storage-blob feel awkward since so much info is populated from headers, and not just from HEAD requests. 

I'd understand if you needed to use the callback to get headers that aren't modeled, but otherwise think I shouldn't need to use a callback to get essential information (that should be part of the response resolved by the promise!)

The goal is to be able to get rid of `_response` right? Removing modeled data from the response seems like a step too far.

Reply:
I know that I am making a conservative assumption here that we can not know just from the swagger whether a certain header is critical for implementing the operation in the SDK. But I believe this is needed to get around the issue of trying to attach stuff to values deserialized as `null`/`undefined`.

>  It would be pretty frustrating for users to fish out etag using the callback, so I'd expect it to be added to the return value somehow. 

The idea is to have the call back mechanism in the generated code but the convenience layer can use that to provide an ideal experience to our customers, in this case attaching `etag` if the return value is not `null`/`undefined`. There is nothing that stops us from even having `_response` back in the convenience layer using this setup.

> My take on this is that if header values are important enough to document in swagger as part of the response, you should attach them to the default result.

Note that default values are not always provided.

> Does this change make it so we can no longer access fields that are populated from headers even if they are modeled in OpenAPI? That would be very annoying as a user and API author.

As an API author, why would you think sending a callback to access the header values needed in implementing your method is annoying? I think it is a simple task.

Reply:
> But I believe this is needed to get around the issue of trying to attach stuff to values deserialized as null/undefined.

Are you referring to the scenario where an operation's body is modeled as an array or scalar value (i.e. not an object) and the service responds with null/undefined? I wonder if that's actually valid, or if they should always be returning a value in this case (e.g. empty array, true or false boolean value, etc.)

If these values can actually be null/undefined, have we considered wrapping these types of operations such that the body is a member of an object instead?

> As an API author, why would you think sending a callback to access the header values needed in implementing your method is annoying? I think it is a simple task.

Take the Storage Blob GetAccountInfo operation as an example. This is a _GET_ operation with _all_ of its data returned as headers. The service has specified what the client name (via `x-ms-client-name`) should appear as, and a description for most of the fields that will then show up in documentation.

If the headers aren't put in the response object, instead of seeing something like:
```js
const {skuName, accountKind} = await client.getAccountInfo();
console.log(skuName, accountKind);
```
They'll need to pass a callback to get the same information:

```js
let skuName;
let accountKind;
await client.getAccountInfo({
  rawResponseCallback(response, rawResponse) {
    skuName = rawResponse.parsedHeaders['skuName'];
    accountKind = rawResponse.parsedHeaders['accountKind'];
  }
});
console.log(skuName, accountKind);
```
It's not clear to me looking at this PR that the user would get type checking support when using the rawResponseCallback either. Will TS complain if I try accessing a parsed header that doesn't exist because I made a typo? Either way, this does not feel like a good user experience compared to what we have today.

>>It would be pretty frustrating for users to fish out etag using the callback, so I'd expect it to be added to the return value somehow.

>The idea is to have the call back mechanism in the generated code but the convenience layer can use that to provide an ideal experience to our customers, in this case attaching etag if the return value is not null/undefined. There is nothing that stops us from even having _response back in the convenience layer using this setup.

I'd really rather we be able to support this scenario in code generation than have to hand-author methods at the convenience layer if possible.

Reply:
@chradek 

>  I wonder if that's actually valid, or if they should always be returning a value in this case (e.g. empty array, true or false boolean value, etc.)

[I have posted on this issue on our teams channel with pointers to such cases in our test server ](https://teams.microsoft.com/l/message/19:344f6b5b36ba414daa15473942c7477b@thread.skype/1607015855831?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=3e17dcb0-4257-4a30-b843-77f47f1d4121&parentMessageId=1607015855831&teamName=Azure%20SDK&channelName=Language%20%E2%80%93%20JS%E2%80%89%EF%BC%86%E2%80%89TS%20%F0%9F%90%B1%E2%80%8D%F0%9F%91%A4&createdTime=1607015855831)

> They'll need to pass a callback to get the same information:

That is true which is pretty simple. You basically need to just have the headers/response object filled by the callback in the convenience layer on-demand.

> It's not clear to me looking at this PR that the user would get type checking support when using the rawResponseCallback either. Will TS complain if I try accessing a parsed header that doesn't exist because I made a typo? Either way, this does not feel like a good user experience compared to what we have today.

This change shifts the authority of specifying which headers to expose to our users to the library author. At the end of the day, the library author is the sole architect of the user experience since they have full control in the convenience layer to expose whatever types to the user in whatever form.

Reply:
I don't believe the raw response callback is intended for any mainline scenario. If it shows up in the README as a snippet for a common operation, I think that's evidence that we have failed at accomplishing our goal of being **approachable**. It's only supposed to be necessary for advanced cases where a consumer needs to debug the raw payload.

While our packages today are mostly based on REST over HTTP, I think that for a large number of consumers, the minutiae of HTTP is below their abstraction level. They don't care if the library is talking over HTTP or gRPC or GraphQL. All they care about is they call a method and get back a typed object that describes a server-side entity.

With that in mind, the distinction of data being returned as headers rather than the HTTP body is not meaningful. All that matters is which parts of the response payload are valuable for most use cases. These are described by Swagger/OpenAPI to help us deserialize them correctly and consistently.

This helper method's purpose is to take the parsed headers and body objects and flatten them into a response object that matches the shape described by the Swagger/OpenAPI file.

There are only really three cases here:

1. The response body is a JS array/object
2. No headers are described (headersMapper is empty on `OperationSpec`) and the response is a JS primitive type (number, string)
3. Some parts of the response are populated from headers and the response is a JS primitive type (number, string)

In the first case, we can freely append header values if they are given by the `OperationSpec` and there is no trouble.
In the second case, we can return the raw value because there are no headers to worry about.
The third case is the only trouble spot, and my recommendation is to detect it and do something like `{ ...parsedHeaders, value: responseValue}`. It's simple, it's predictable, and the convenience client can throw away the headers and return only the value property if it wants.

Reply:
@xirzec 

> I don't believe the raw response callback is intended for any mainline scenario. If it shows up in the README as a snippet for a common operation, I think that's evidence that we have failed at accomplishing our goal of being approachable. It's only supposed to be necessary for advanced cases where a consumer needs to debug the raw payload.

Sure! This burden is on us though to make sure we provide a nice user experience.

I am not sure where standard protocol headers fit in your proposal though. Are you suggesting a hybrid design combining the existing one (for OperationSpec headers) with the callback one (for standard headers)?

Reply:
> No headers are described (headersMapper is empty on OperationSpec) and the response is a JS primitive type (number, string)

@xirzec 
I was talking to @richardpark-msft about this and I think it is worth considering also wrapping the primitive in an object like your proposed solution for option 3. The reason being that a service could decide to add a header sometime in the future, and from a REST perspective that wouldn't be a breaking change. However from a client API perspective it would be since we'd be moving from the 2nd scenario to the 3rd one.

> Sure! This burden is on us though to make sure we provide a nice user experience.

I just want to make sure we're on the same page. I think one of the goals for autorest is to reduce the amount of logic we need to write by hand in the convenience layer. (Ultimately it'd be great if we could have clients fully code-generated!)

With this change, we're going to greatly increase the amount of code we have to write by hand...basically any operation that models response data in headers. I've been focusing on storage blob as an example where this will cause a lot of code to need to be written where it previously didn't need to be. And I'm not sure what the benefit is for our users...it sounds like we're saying they'll have no change in experience. I don't think this is a solution that would scale long term.

I'd be interested in a real-world scenario where a service returns a primitive type as the response without specifying headers...the teams thread you posted only mentions the testserver swaggers which aren't necessarily what we'd see in the wild. Is this actually a common scenario? More common than having headers modeled? That said, I echo @xirzec's sentiment that the convenience layer can always discard the extra information (e.g. headers) if they aren't needed if we continue wrapping primitives in an object, and I suspect this scenario would occur much less frequently than having to add modeled headers.

Reply:
I agree with @chradek that we should always wrap primitive values to sidestep future compatibility concerns. 

Which actually I think brings us all the way back around to what the code does today since it uses `isPrimitiveType(fullResponse.parsedBody)` to see if it should wrap the body with an object and give it the property name `body`.

So if I'm not mistaken, the ask here is simply to back out all changes except the removal of `_response`

Reply:
> So if I'm not mistaken, the ask here is simply to back out all changes except the removal of _response

Which, weren't we only removing `_response` due to not being able to add it onto `null` or `undefined`? If we always wrap primitives, does it still make sense to make this change? Maybe there are other reasons to remove it as well (potentially better memory usage since we're not holding onto everything in the response?).

@richardpark-msft made a good point when we chatted earlier that using the callback can be tricky if you're not awaiting the operation and instead using the promise (maybe in a Promise.all() call, or passing it to another method).

Reply:
To me, the main motivations of removing `_response` are:

1. Trying to type it in generated code is ugly and messy (as it creates an additional wrapping type for every response object.)
2. Often in convenience we forget to re-attach it to converted models (or attach it incorrectly as enumerable instead of non-enumerable.)
3. What to do in the case of an operation having multiple responses is ambiguous (this happened in data-tables I believe), but we could just call the callback multiple times.
4. In the cases where a convenience layer **would** like to return a simple value (e.g. the count operation in a search index) it would be nice to not worry about having to make a wrapper just to include `_response` which would rarely be needed.

Reply:
> 1. Trying to type it in generated code is ugly and messy (as it creates an additional wrapping type for every response object.)

Would something like `YourObject & EmbeddedResponseType` allow us to "mixin" the _response property while still providing clarity? The _response here would have the same shape for everyone.

This might also reveal my misunderstanding - in this case the customer is _us_ internally, right? So having the raw and untyped (but consistent) _response should be fine.

> 3. What to do in the case of an operation having multiple responses is ambiguous (this happened in data-tables I believe), but we could just call the callback multiple times.

This one is interesting. Would the user/caller even think that the call will happen multiple times for a single request? 

Regardless of approach taken this seems like an interesting complication and, I think, brings @chradek's concern back to the fore. It seems more complex to deal with multiple calls to a callback (and having to write code to aggregate it together) vs something simple like just having an array.

> 4. In the cases where a convenience layer would like to return a simple value (e.g. the count operation in a search index) it would be nice to not worry about having to make a wrapper just to include _response which would rarely be needed.

Returning the primitive directly from the convenience layer is bad from an API compatibility standpoint, right? How would you "upgrade" that later on if there was additional data to return from that call? Coming back to the REST standpoint - they could be backwards compatible in the underlying REST endpoint by simply adding headers but we'd have no backwards compatible way to represent that in our convenience layer.

(apologies if this entire thread of discussion is already settled)

Reply:
I think this was resolved by #13132